File: codeforces/SubstringGraph.java
import java.util.*;
import java.io.*;
import java.math.*;
public class SubstringGraph{
    static HashMap<Integer, ArrayList<Integer>> adj=new HashMap<>();
    static ArrayList<Integer> topo=new ArrayList<>();
    static int dp[][];
    static int mark[];
    public static boolean cycle(int v){
        if(mark[v]==2) return false;
        if(mark[v]==1) return true;
        mark[v]=1;
        for(int i=0;i<adj.get(v).size();i++){
            if(cycle(adj.get(v).get(i)))
                return true;
        }
        mark[v]=2;
        topo.add(v);
        return false;
    }
    public static void main(String[] args) {
        MyScanner sc=new MyScanner();
        int n=sc.nextInt();
        int m=sc.nextInt();
        String s=sc.next();
        s="%"+s;
        mark=new int[n+7];
        for(int i=1;i<=n;i++) adj.put(i,new ArrayList<Integer>());
        for(int i=0;i<m;i++){
            int v1=sc.nextInt();
            int v2=sc.nextInt();
            adj.get(v1).add(v2);
        }
        //cycle detection
        for(int i=1;i<=n;i++){
            if(cycle(i)){System.out.println(-1);System.exit(0);}
        }
        //topo check
        // for(Integer x:topo)
        //     System.out.println("topo:"+x);

        int answer=0;
        dp=new int[26][n+7];
        for(int i='a';i<='z';i++){
            for(int j=0;j<topo.size();j++){
                int v=topo.get(j);
                int val=0;
                if(s.charAt(v)==(char)i)
                    val=1;
                dp[i-'a'][v]=val;
                for(int k=0;k<adj.get(v).size();k++){
                    int neigh=adj.get(v).get(k);
                    dp[i-'a'][v]=Math.max(dp[i-'a'][v],dp[i-'a'][neigh]+val);
                } // end of k
                answer=Math.max(answer,dp[i-'a'][v]);
            }
        }
        System.out.println(answer);
    }


    private static class MyScanner {
      BufferedReader br;
      StringTokenizer st;

      public MyScanner() {
         br = new BufferedReader(new InputStreamReader(System.in));
      }

      String next() {
          while (st == null || !st.hasMoreElements()) {
              try {
                  st = new StringTokenizer(br.readLine());
              } catch (IOException e) {
                  e.printStackTrace();
              }
          }
          return st.nextToken();
      }

      int nextInt() {
          return Integer.parseInt(next());
      }

      long nextLong() {
          return Long.parseLong(next());
      }

      double nextDouble() {
          return Double.parseDouble(next());
      }

      String nextLine(){
          String str = "";
	  try {
	     str = br.readLine();
	  } catch (IOException e) {
	     e.printStackTrace();
	  }
	  return str;
      }
   }
}

File: codeforces/SubstringGraph.cpp
#include <cstdio>
#include <iostream>
#include <vector>
#include <set>
#include <map>
#include <cmath>
#include <string>
#include <cstring>
#include <sstream>
#include <queue>
#include <iomanip>
#include <algorithm>
using namespace std;
const int mx=300007;
static int mark[mx];
static int dp[26][mx];
vector<int> adj[mx];
vector<int> topo;

bool cycle(int v){
	if(mark[v]==2) return false;
	if(mark[v]==1) return true;
	mark[v]=1;
	for(int i=0;i<adj[v].size();i++)
		if(cycle(adj[v][i])) return true;
	mark[v]=2;
	topo.push_back(v);
	return false;
}
int main(){
	ios_base::sync_with_stdio(0); cin.tie(0);cout.tie(0);
	int n,m;
	cin>>n>>m;
	string s;
	cin>>s;
	s="%"+s;
	for(int i=0;i<m;i++){
		int x,y;
		cin>>x>>y;
		adj[x].push_back(y);
	}
	//cycle found
	for(int i=1;i<=n;i++){
		if(cycle(i)){cout<<"-1"<<endl;return 0;}
	}
	int answer=0;
	for(int i='a';i<='z';i++){
		for(int j=0;j<topo.size();j++){
			int v=topo[j];
			int val=0;
			if(i==s[v])
				val=1;
			dp[i-'a'][v]=val;
			for(int k=0;k<adj[v].size();k++)
				dp[i-'a'][v]=max(dp[i-'a'][v], dp[i-'a'][adj[v][k]]+val);
			answer=max(answer,dp[i-'a'][v]);
		}//end of j
	}//end of i
	cout << answer << endl;
	return 0;
}
